### PR TITLE
Harden elasticsearch chart for Kube 1.5

### PR DESCRIPTION
**Pull request by icereval**
Further extends upon PR https://github.com/kubernetes/charts/pull/890

* Added configmap to explicitly provide cluster configurations and scripts

* Replace depreciating `ES_HEAP_SIZE` with `ES_JAVA_OPTS` to position for ES v5 support

* Removed alpha storage class operators

* Removed catastrophic liveness probe checking entire clusters health

* Readiness probe now inspects local node health

* Added termination grace period (defaults to 60m) to allow pre-stop-script.sh time to gracefully migrate shards

* Added init container to configure `vm.max_map_count`

* Updated elasticsearch.yaml:
  * Added `PROCESSOR` configuration to prevent large cluster garbage collection issues leading to node eviction
  * Added configurable gateway defaults to help avoid a split brain, requiring two masters online and in consensus before recovery can continue

* Updated pre-stop-script.sh:
  * Check `v1beta1` `statefulset` endpoint
  * Evalute `.spec.replicas` for statefulset desired size
  * Clear `_cluster/settings` ip exclusion prior to shutdown to avoid a possible (random) ip match scenario on expansion of the clsuter

* Data nodes now use default storage class if one is not specified

* Apply Helm best practices to prep for stable

### CODE REVIEW

File: stable/elasticsearch/values.yaml [hunk 1]
```diff
@@ -29,5 +29,5 @@ DataCpuLimits: "1"
 DataMemoryLimits: "4Gi"
 DataHeapSize: "1536m"
 DataStorage: "30Gi"
-DataStorageClass: "anything"
-DataStorageClassVersion: "alpha"
+# DataStorageClass: "ssd"
+DataTerminationGracePeriodSeconds: 900
```
Comment (unguiculus) [inline stable/elasticsearch/values.yaml]: I would prefer to not move the chart to `stable` as long as config values start with an uppercase letter. In this case, I would also switch to nesting first. I think this make sense here, e. g.:

```yaml
client:
  replicas: 2
  port: 2379
  heapSize: "128m"
master:
  replicas: 2
  heapSize: "128m"
data:
  replicas: 3
  storage: "30Gi"
```

See https://github.com/kubernetes/helm/blob/master/docs/chart_best_practices/values.md

Generally, I think a move to stable should be a separate PR so it is explicit.

Regarding resources, it's becoming increasingly popular to specify resources like so (without specifying defaults): https://github.com/kubernetes/charts/blob/master/stable/nginx-ingress/templates/controller-deployment.yaml#L93

**Review by unguiculus — CHANGES_REQUESTED** [review]

File: stable/elasticsearch/values.yaml (hunk 1 above)
Comment (icereval) [reply stable/elasticsearch/values.yaml]: I completely agree and all are very good points. Would you mind informing the base PR https://github.com/kubernetes/charts/pull/890, if we can get this merged, even if still in incubator status one of us can follow up with a new PR and the final changes needed to make this `stable`.

**Comment by viglesiasce** [comment]
@simonswine @unguiculus any update on this review? Were the latest changes enough to merge?

**Comment by unguiculus** [comment]
As suggested in #890, a move to stable should only happen after applying best practices, e. g. regarding values. See https://github.com/kubernetes/helm/blob/master/docs/chart_best_practices/values.md.

**Comment by icereval** [comment]
@unguiculus & @prydonius, rebase complete and a first pass at helm best practices applied to the incubator chart.

**Comment by unguiculus** [comment]
Excellent. This goes in the right direction. Now, please have a look at `app` labels, which should be `{{ template "name" . }}`. Note that you will then have to add the `release` label to the selector in services. `nginx-ingress` an excellent example:

https://github.com/kubernetes/charts/blob/master/stable/nginx-ingress/templates/controller-deployment.yaml#L6
https://github.com/kubernetes/charts/blob/master/stable/nginx-ingress/templates/controller-service.yaml#L56-L58

**Comment by unguiculus** [comment]
@icereval Are you working on an update?

**Comment by icereval** [comment]
@unguiculus, yep, I'll have my updates ready soon

**Comment by icereval** [comment]
@unguiculus, best practice changes pushed up

**Comment by unguiculus** [comment]
Would you mind adding a NOTES.txt? Otherwise it looks nice but I have yet to review more thoroughly. I installed it on GKE and everything came up nicely.

**Comment by unguiculus** [comment]
@k8s-bot ok to test

**Comment by icereval** [comment]
Added NOTES.txt and `client.serviceType` based on review of the concourse stable chart.

**Review by unguiculus — APPROVED** [review]

**Pull request merged**